python/eos/tasks.py

```diff
@@ -0,0 +1,196 @@
+    :param pre_N: The number of samples to be used for an adaptation in each prerun steps. These samples will be discarded.
+    :type pre_N: int, optional
+    :param preruns: The number of prerun steps, which ared used to adapt the MCMC proposal to the posterior.
+    :type preruns: int, optional
+    :param N: The number of samples to be stored in the output file. Defaults to 1000.
+    :type N: int, optional
+    :param stride: The ratio of samples drawn over samples stored. For every S samples, S - 1 will be discarded. Defaults to 5.
+    :type stride: int, optional
+    """
+
+    output_path = os.path.join(base_directory, posterior, 'mcmc-{:04}'.format(int(chain)))
+    _set_log_file(output_path, 'log')
+    _analysis_file = eos.AnalysisFile(analysis_file)
+    analysis = _analysis_file.analysis(posterior)
+    rng = _np.random.mtrand.RandomState(int(chain) + 1701)
+    try:
+        samples, weights = analysis.sample(N=N, stride=stride, pre_N=pre_N, preruns=preruns, rng=rng)
+        eos.data.MarkovChain.create(output_path, analysis.varied_parameters, samples, weights)
+    except RuntimeError as e:
+        eos.error('encountered run time error ({e}) in parameter point:'.format(e=e))
+        for p in analysis.varied_parameters:
+            eos.error(' - {n}: {v}'.format(n=p.name(), v=p.evaluate()))
+
+def find_clusters(analysis_file, posterior, base_directory='./', threshold=2.0, K_g=1):
+    """
+    Finds clusters among posterior MCMC samples, grouped by Gelman-Rubin R value, and creates a Gaussian mixture density.
+
+    Finding clusters and creating a Gaussian mixture density is a neccessary intermediate step before using the sample-pmc subcommand.
+    The input files are expected in EOS_BASE_DIRECTORY/POSTERIOR/mcmc-*. All MCMC input files present will be used in the clustering.
+    The output files will be stored in EOS_BASE_DIRECTORY/POSTERIOR/clusters.
+
+    :param analysis_file: The name of the analysis file that describes the named posterior.
+    :type analysis_file: str
+    :param posterior: The name of the posterior.
+    :type posterior: str
+    :param base_directory: The base directory for the storage of data files. Can also be set via the EOS_BASE_DIRECTORY environment variable.
+    :type base_directory: str, optional
+    :param threshold: The R value threshold. If two sample subsets have an R value larger than this threshold, they will be treated as two distinct clusters. Defaults to 2.0.
+    :type threshold: float > 1.0, optional
+    :param K_g: The number of mixture components per cluster. Default to 1.
+    :type K_g: int >= 1, optional
+    """
+
+    import pathlib
+    output_path = os.path.join(base_directory, posterior, 'clusters')
+    input_paths = [str(p) for p in pathlib.Path(os.path.join(base_directory, posterior)).glob('mcmc-*')]
+    chains    = [eos.data.MarkovChain(path).samples for path in input_paths]
+    means     = [_np.mean(chain, axis=0) for chain in chains]
```

Comment:
`_np.mean(chains, axis=0)` should give the same result in a more efficient way, I think

Reply:
Hm, isn't it ``_np.mean(chains, axis=1)``? That seems to work for me. Changed.

Reply:
Yes you are right!